build/build-viewer.js

```diff
@@ -0,0 +1,172 @@
+    'sw.js',
+    'manifest.json',
+  ], distDir, {
+    cwd: `${sourceDir}/app/`,
+    parents: true,
+  });
+
+  // Copy polyfills.
+  return cpy([
+    '../node_modules/url-search-params/build/url-search-params.js',
+    '../node_modules/whatwg-fetch/fetch.js',
+  ], `${distDir}/src/polyfills`, {cwd: sourceDir});
+}
+
+/**
+ * Concat report and viewer stylesheets into single viewer.css file.
+ * @return {Promise<void>}
+ */
+async function css() {
+  const reportCss = htmlReportAssets.REPORT_CSS;
+  const viewerCss = await readFileAsync(`${sourceDir}/app/styles/viewer.css`, {encoding: 'utf8'});
+  await safeWriteFileAsync(`${distDir}/styles/viewer.css`, [reportCss, viewerCss].join('\n'));
+}
+
+/**
+ * Insert report templates into html and copy to dist.
+ * @return {Promise<void>}
+ */
+async function html() {
+  let htmlSrc = await readFileAsync(`${sourceDir}/app/index.html`, {encoding: 'utf8'});
+  htmlSrc = htmlSrc.replace(/%%LIGHTHOUSE_TEMPLATES%%/, htmlReportAssets.REPORT_TEMPLATES);
+
+  await safeWriteFileAsync(`${distDir}/index.html`, htmlSrc);
+}
+
+/**
+ * Combine multiple JS files into single viewer.js file.
+ * @return {Promise<void>}
+ */
+async function compileJs() {
+  // JS bundle from browserified ReportGenerator.
+  const generatorFilename = `${sourceDir}/../lighthouse-core/report/report-generator.js`;
+  const generatorBrowserify = browserify(generatorFilename, {standalone: 'ReportGenerator'})
+    .transform('brfs');
+  const generatorBundle = promisify(generatorBrowserify.bundle.bind(generatorBrowserify));
+  const generatorJs = (await generatorBundle()).toString();
+
+  // Report renderer scripts.
+  const rendererJs = htmlReportAssets.REPORT_JAVASCRIPT;
+
+  // idb-keyval dependency.
+  const idbKeyvalPath = require.resolve('idb-keyval/dist/idb-keyval-min.js');
+  const idbKeyvalJs = await readFileAsync(idbKeyvalPath, 'utf8');
+
+  // Current Lighthouse version as a global variable.
+  const versionJs = `window.LH_CURRENT_VERSION = '${lighthousePackage.version}';`;
+
+  // Viewer-specific JS files.
+  const viewJsFiles = await loadFiles(`${sourceDir}/app/src/*.js`);
+
+  const contents = [
+    generatorJs,
+    rendererJs,
+    idbKeyvalJs,
+    versionJs,
+    ...viewJsFiles,
+  ];
+  const options = {
+    output: {preamble: license}, // Insert license at top.
+  };
+  const uglified = uglifyEs.minify(contents, options);
+  if (uglified.error) {
+    throw uglified.error;
+  }
+
+  await safeWriteFileAsync(`${distDir}/src/viewer.js`, uglified.code);
+}
+
+/**
+ * Build viewer, optionally deploying to gh-pages if `--deploy` flag was set.
+ */
+async function run() {
+  // Clean and build.
+  rimraf.sync(distDir);
+  await Promise.all([
+    compileJs(),
+    html(),
+    css(),
+    copyAssets(),
+  ]);
+
+  const argv = process.argv.slice(2);
+  if (argv.includes('--deploy')) {
+    await ghPages(`${distDir}/**/*`, {
+      add: true, // keep existing files
+      dest: 'viewer',
+    }, () => {});
+  }
+}
+
+run();
```

Comment:
I was about to write something along the lines of shame we have to have all this code, but then I saw that it's even fewer lines than the gulpfile 😆